[WIP] Added Llama-vid

nileshkokane01 opened the pull request:
# What does this PR do?

Adds Llama-VID 


<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #27980

For video: 
```python
from PIL import Image
import requests
from transformers import LLaMAVIDLlavaForConditionalGeneration, LLaMAVIDLlavaProcessor,LLaMAVIDLlavaConfig
from huggingface_hub import hf_hub_download
import torch
import numpy as np
from decord import VideoReader, cpu



def load_video(video_path, fps=1):
    vr = VideoReader(video_path, ctx=cpu(0))
    fps = round(vr.get_avg_fps()/fps)
    frame_idx = [i for i in range(0, len(vr), fps)]
    spare_frames = vr.get_batch(frame_idx).asnumpy()
    return spare_frames

model = LLaMAVIDLlavaForConditionalGeneration.from_pretrained("Nilesh360/llama-vid-7b-full-224-video-fps-1")
processor = LLaMAVIDLlavaProcessor.from_pretrained("Nilesh360/llama-vid-7b-full-224-video-fps-1")

prompt = "<image>\nUSER: What's the content of the video?\nASSISTANT:"
video= hf_hub_download(repo_id="nielsr/video-demo", filename="eating_spaghetti.mp4", repo_type="dataset")
images= load_video(video)

inputs = processor(text=[prompt], images=[images], return_tensors="pt")
# Generate
generate_ids = model.generate(**inputs, max_length=512)
reply = processor.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0]
print(reply)

```
For image : 
```python
from PIL import Image
import requests
from transformers import LLaMAVIDLlavaForConditionalGeneration, LLaMAVIDLlavaProcessor,LLaMAVIDLlavaConfig
import torch
import numpy as np

processor = LLaMAVIDLlavaProcessor.from_pretrained("Nilesh360/llama-vid-7b-full-336")
model = LLaMAVIDLlavaForConditionalGeneration.from_pretrained("Nilesh360/llama-vid-7b-full-336")

prompt = "<image>\nUSER: What's the content of the image?\nASSISTANT:"
url= 'https://huggingface.co/datasets/huggingface/cats-image'
image=  Image.open(requests.get(url, stream=True).convert('RGB'))


inputs = processor(text=[prompt], images=[image], return_tensors="pt")
# Generate
generate_ids = model.generate(**inputs, max_length=512)
reply = processor.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0]
print(reply)
```

Colab: https://colab.research.google.com/drive/1fvaagb4nSjtEvx2YDrcqY0E2jWIsdCtv#scrollTo=v4Ry7RyyiruU



## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?
@amyeroberts

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

nileshkokane01 committed:
Added Llama-vid draft

nileshkokane01 committed:
Deleted irrelevant files and added few relevent files

amyeroberts commented:
Awesome work @nileshkokane01 Thanks for opening a PR! Let us know when it's ready for review or you have any questions about the port into transformers

amyeroberts commented:
@nileshkokane01 Yep, that's the right approach. Ideally we can leverage existing architectures as much as possible using `# Copied from` comments. I'd suggest looking at llava as well as llama.

nileshkokane01 committed:
added base from llava

nileshkokane01 commented:
@amyeroberts ,

Based on the [comment](https://github.com/dvlab-research/LLaMA-VID/issues/13#issuecomment-1855744892) from the author, what I understand is [VisionTransformer](https://github.com/dvlab-research/LLaMA-VID/blob/main/llamavid/model/multimodal_encoder/builder.py#L16) should be replaced by [BLIP2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/blip_2/modeling_blip_2.py) and the [qformer](https://github.com/dvlab-research/LLaMA-VID/blob/main/llamavid/model/llamavid_arch.py#L104) with [instructBLIP](https://github.com/huggingface/transformers/blob/main/src/transformers/models/instructblip/modeling_instructblip.py). 

Can you please confirm ?

nileshkokane01 commented:
comparing, BLIP2 and instructBLIP visionModels, they turned out to be same, so I use the instructBLIP for VIT and Qformer.

nileshkokane01 committed:
Added basic functionality for LLaMaVID that includes:

1. Conversion  of model to hf.
2. Utilized InstructBLIP qformer and Vision encoder for geneartion of embeddings.
3. Post processing of qformer output to input of LLama model.
4. Few other nits.

amyeroberts commented:
> That also means I have to copy the qformer and the associated embedding class etc from the instructBLIP with #copied from comments ?

Yep!

nileshkokane01 committed:
Modified code to make it consistant with the original implementation to precision 1e3 as per the doc: https://huggingface.co/docs/transformers/en/add_new_model

nileshkokane01 committed:
Forward passed for video and tested the output.

nileshkokane01 committed:
Cleaned code and added #Copied from whereever necessery.

nileshkokane01 committed:
Make fixup, style , and quality

nileshkokane01 committed:
Clean-up and some refactoring of code

nileshkokane01 committed:
Few test cases added as dummy placeholders

nileshkokane01 committed:
fixed nits

NielsRogge commented:
cc @zucchini-nlp as she is currently working on adding Video-LLaVa

nileshkokane01 commented:
@amyeroberts @NielsRogge ,

Any hints, I am really interested getting this merged. Please let me know.

amyeroberts commented on src/transformers/models/llamavid/modeling_llamavid.py:
```diff
@@ -0,0 +1,1906 @@
+        return layer_output
+
+
+# Copied from transformers.models.blip_2.modeling_blip_2.Blip2QFormerEncoder with Blip2->InstructBlip
```
Comment here would rename to instructblip - but you want to rename to `LLaMAVIDLlava`

amyeroberts commented on src/transformers/models/llamavid/modeling_llamavid.py:
```diff
@@ -0,0 +1,1906 @@
+        super().__init__(config)
+        self.config = config
+        self.vision_tower = LLaMAVIDLlavaVisionModel(config.vision_tower_config)
+        self.qformer = LLaMAVIDLlavaQFormerModel(config.qformer_config)
```
If all of this is copied from, we can use AutoModel instead

amyeroberts commented on src/transformers/models/llamavid/modeling_llamavid.py:
```diff
@@ -0,0 +1,1906 @@
+            block_size = 64
+            outputs = []
+            ctx_score = ctx_embed.softmax(-1)
+            for L in range(0, len(ctx_score), block_size):
+                R = L + block_size
+                sub_embed = (ctx_score[L:R] @ vis_embed[L:R]).mean(1)
```
Convention is not single letter var names

amyeroberts reviewed (commented):
Thanks for all the work adding this model! 

I've made a first high-level review to give some pointers on the structure. 

In terms of similarity between the original and ported model, the expected tolerance is 1e-5, rather than 1e-3 and should really be tested on GPUs. This will also be necessary for adding the integration tests for the models. Have you tried running with GPUs on colab or paperspace?

qgallouedec commented:
Thanks for the hard work here. This PR gone stale, so I'm closing it. More here https://github.com/huggingface/transformers/issues/27980#issuecomment-5085930041
